🚀 3단계 - 자동차 경주

giyeon95 opened the pull request:
리뷰어님 안녕하세요!
3단계 미션 리뷰 요청 드립니다.🙏

감사합니다!

giyeon95 committed:
feat: 🚀 3단계 - 자동차 경주 요구사항 정리

giyeon95 committed:
feat: 자동차 도메인(Car) 기능 구현

giyeon95 committed:
feat: 각 round별 상태를 기록하는 (RaceStage) 기능 구현

giyeon95 committed:
feat: 자동차 갯수 및 라운드 입력 및 결과 출력 기능 구현

giyeon95 committed:
refactor: 매직넘버 제거

ddaaac commented on src/main/kotlin/racingcar/Car.kt:
```diff
@@ -0,0 +1,23 @@
+    val pos: Int = 0
+) {
+
+    fun race(rnd: Random): Car {
+        return when (rnd.nextInt(LIMIT)) {
+            in START until FORWARD -> Car(pos)
+            in FORWARD until LIMIT -> Car(pos + MOVE_POS)
+            else -> throw IllegalArgumentException("System error")
+        }
+    }
```
테스트 가능한 구조로 함수를 만들어주셨네요 또, 코틀린의 적절한 확장함수와 키워드를 사용해주셨구요. 💯

- 지금은 코틀린에서 제공하는 `Random`을 파라미터로 받고 있습니다. 그러면서 불필요한 오버라이딩이 발생하고 있는 것 같아요. 직접 인터페이스를 정의해도 되지 않을까 싶네요.
- 전진을 하지 않을 때는 this를 반환해도 괜찮을 것 같아요.

giyeon95 replied:
첫번째 말씀주신 부분에 대하여 질문이 있는데요!

1. 코틀린에서 제공하는 Random을 파라미터로 받고있어 불필요한 오버라이딩이 발생하고 있는 내용은 테스트 코드 내  NextRandomFixture `nextBits(bitCount)` 을 말씀주신 게 맞을까요?
2. 직접 인터페이스를 정의하는 말씀이 정확하게 이해가 되지 않는데요, 말씀주신 직접 정의하는 인터페이스가 전진할지, 유지할지 상태(?)를 표기하는 인터페이스를 말씀하신 것일까요? 혹은 약간의 예시를 받아볼 수 있을까요?

ddaaac replied:
1. 넵 픽스쳐 부분에서 불필요하게 override가 발생하는 것 같아서 드린 코멘트였습니다.
2. 크게 두가지로 작성할 수 있을 것 같아요.

```kotlin
interface RacingCarRandom {
    fun nextInt(): Int
}

interface RacingCarEngine {
    fun canMove(): Boolean
}
```

어떤 방법이던 레이싱 카 도메인에 적합한 인터페이스를 도출해볼 수 있지 않을까요?

ddaaac commented on src/main/kotlin/racingcar/Car.kt:
```diff
@@ -0,0 +1,23 @@
+
+private const val MOVE_POS = 1
+
+class Car(
+    val pos: Int = 0
+) {
```
저는 객체에 초기값을 세팅해줄 때, 빈(기본) 생성자를 이용하는 방식보다는 정적 팩토리 메서드를 선호하는 편입니다. `Position()`와 `Position.initial()` 정도 차이겠네요.
이유는 빈 생성자는 사람들마다 그 뜻을 다르게 해석할 수 있다고 생각해서에요. 누군가에게는 빈 생성자가 당연히 특정 의미로 받아질 수 있겠지만 다른 사람은 다르게 받아들이는 경우를 종종 보았거든요.

대표적인 예시로는 `Date()` 와 `LocalDateTime.now()`의 초기화 방식 차이가 있겠네요. 기연님의 생각은 어떠신가요?

giyeon95 replied:
음 저는.. 정적 팩토리 메서드로 객체의 초기화를 지원해준다면, 객체를 초기화 하는 케이스가 많아져서 `의도를 나타내야 하는 경우`와, `기본 객체 생성을 private으로 캡슐화` 후 제공 해야 이점을 얻어갈 수 있다고 생각이 들어요.😁
> 예시로 제공해주신 LocalDateTime 의 정적 팩토리 메서드도 생성 방식이 많고, 기본 생성자로는 생성이 막혀있으니 좋은 예시인 것 같습니다!


다만 kotlin 으로 구현하게 된다면 `companion object block` 의 추가와, `private constructor` 로 변경을 해주어야 할것 같은데요,
오히려 가독성이 떨어지지 않나? 라는 생각과 불필요한 코드가 많아지지 않나? 라는 생각이 들었습니다. 그렇다보니 단순 객체의 초기화 정도라면, 기본 생성자를 이용해도 되지 않을까? 라는 생각이 들기도 해요.🤔 (초기값으로 지정해주는 케이스가 늘어난다면 정적 팩토리 메서드로 리펙토링 하는 것도 고민해 볼 수 있을것 같아요.)


다만 팀 내에서 `정적 팩토리 메서드를 이용해서 객체를 초기화 하자!` 는 컨벤션과 팀원들과 논의가 되어있으면 팀에 규칙에 따르는게 맞는 것 같다는 의견입니다.😁

ddaaac replied:
좋은 의견 감사합니다!
팀 컨벤션이 있다면 저도 컨벤션을 따를 것 같아요. 
사실 정답이 있는 문제는 아니라 제가 리뷰하는 분들에게 모두 남기는 코멘트인데, 기연님의 생각을 잘 전해주신 것 같아서 좋네요 👍

ddaaac commented on src/main/kotlin/racingcar/RaceMain.kt:
```diff
@@ -0,0 +1,19 @@
+        raceStage = raceStage.racing()
+
+        ResultView.printResult(raceStage)
+    }
```
여기도 `repeat`을 활용할 수 있을 것 같아요!

ddaaac commented on src/main/kotlin/racingcar/RaceStage.kt:
```diff
@@ -0,0 +1,31 @@
+
+class RaceStage(
+    val cars: List<Car>,
+    val round: Int
```
round는 어디에서 사용하는 프로퍼티일까요?

giyeon95 replied:
각 회차별 기록 & 회차별 점수를 체크(?) 하고자 고려를 했었는데요, 구현하고 나니 필요가 없어진 프로퍼티였네요..ㅎㅎ😅

ddaaac commented on src/main/kotlin/racingcar/RaceStage.kt:
```diff
@@ -0,0 +1,31 @@
+    }
+
+    companion object {
+        fun create(carCount: Int): RaceStage = RaceStage(
+            (0 until carCount).map { Car() }, INIT_ROUND
+        )
```
[코틀린의 리스트 초기화 방식](https://www.baeldung.com/kotlin/initialize-list#list-builders)에는 여러가지 방식이 있습니다. 한 번 둘러보셔도 좋을 것 같아요!

giyeon95 replied:
오, 뭔가 더 깔끔한 방법이 없을까 생각했는데 아래와 같이 작성하면 훨씬 읽기 좋네요👍👍

`List(carCount) { Car() }`

ddaaac commented on src/main/kotlin/racingcar/view/InputView.kt:
```diff
@@ -0,0 +1,18 @@
+package racingcar.view
+
+class InputView private constructor() {
```
싱글톤 패턴을 구현하고 싶으셨던 것 같아요. 코틀린은 이를 위해 [object](https://kotlinlang.org/docs/object-declarations.html#object-declarations-overview) 키워드를 제공합니다 :)

giyeon95 replied:
👍👍👍👍👍👍👍👍👍👍👍👍👍

ddaaac commented on src/main/kotlin/racingcar/view/InputView.kt:
```diff
@@ -0,0 +1,18 @@
+        fun inputCarCount(): Int {
+            println("자동차 대수는 몇 대인가요?")
+
+            return readLine()!!.toInt()
```
```suggestion
            return readln().toInt()
```
코틀린 1.6부터 추가된 새로운 함수입니다!
+) 입력값이 Int가 아닐 경우에 커스텀 예외를 제공하고 싶다면 `toIntOrNull`을 활용할 수 있을 것 같아요. 이처럼 코틀린은 제공하는 함수들에 null/non-null을 리턴하는 함수 두 개를 모두 제공하는 경우가 많습니다. 상황에 맞게 사용해주시면 될 것 같아요.

참고로, non-null assertion 연산자를 !!처럼 불편한 모양으로 정한 건, 사용할 때 한 번 더 생각해보고 불편하게 사용하라는 코틀린 개발진들의 의도가 들어가있다고 합니다 😄 코틀린이 제공하는 기본 함수에 이를 사용하려고 한다면 non-nullable한 리턴값을 갖는 동일한 함수가 있는지 확인해봐도 좋을 것 같아요.

giyeon95 replied:
저만 느끼기에 !!이 불편했던게 아니군요, 좋은 팁 감사합니다.🙏

ddaaac commented on src/main/kotlin/racingcar/view/ResultView.kt:
```diff
@@ -0,0 +1,22 @@
+        }
+
+        fun printResult(raceStage: RaceStage) {
+            val progress = raceStage.cars.joinToString(separator = LINE_BREAK) { car -> SCORE_SYMBOL.repeat(car.pos) }
```
적절하게 확장함수들을 사용해주셨네요 💯

ddaaac commented on src/test/kotlin/racingcar/CarTest.kt:
```diff
@@ -0,0 +1,21 @@
+
+internal class CarTest {
+
+    @CsvSource(value = ["0:0", "1:0", "2:0", "3:0", "4:1", "5:1", "6:1", "7:1", "8:1", "9:1"], delimiter = ':')
```
예외값에 대한 케이스도 정의해 두셨는데, 이에 대한 테스트도 있으면 좋을 것 같아요.

ddaaac reviewed (changes requested):
안녕하세요 기연님 👋 

깔끔하게 구현 잘 해주셨네요 👍 

몇 가지 코멘트를 남겼으니 확인 부탁드려요~

giyeon95 committed:
feat: for-loop -> repeat 으로 변경

giyeon95 committed:
feat: RaceStage 내 round 프로퍼티 제거

round는 어디에서 사용하는 프로퍼티일까요?

_Originally posted by @ddaaac in https://github.com/next-step/kotlin-racingcar/pull/760#discussion_r1015675525_

giyeon95 committed:
feat: 리스트 초기화 방식 코틀린스러운 방식으로 변경

[코틀린의 리스트 초기화 방식](https://www.baeldung.com/kotlin/initialize-list#list-builders)에는 여러가지 방식이 있습니다. 한 번 둘러보셔도 좋을 것 같아요!

_Originally posted by @ddaaac in https://github.com/next-step/kotlin-racingcar/pull/760#discussion_r1015682634_

giyeon95 committed:
refactor: object 키워드를 이용하여 View 클래스 개선

싱글톤 패턴을 구현하고 싶으셨던 것 같아요. 코틀린은 이를 위해 [object](https://kotlinlang.org/docs/object-declarations.html#object-declarations-overview) 키워드를 제공합니다 :)

_Originally posted by @ddaaac in https://github.com/next-step/kotlin-racingcar/pull/760#discussion_r1015684445_

giyeon95 committed:
refactor: readLine()!! -> readln() 으로 개선

코틀린 1.6부터 추가된 새로운 함수입니다!
+) 입력값이 Int가 아닐 경우에 커스텀 예외를 제공하고 싶다면 `toIntOrNull`을 활용할 수 있을 것 같아요. 이처럼 코틀린은 제공하는 함수들에 null/non-null을 리턴하는 함수 두 개를 모두 제공하는 경우가 많습니다. 상황에 맞게 사용해주시면 될 것 같아요.

참고로, non-null assertion 연산자를 !!처럼 불편한 모양으로 정한 건, 사용할 때 한 번 더 생각해보고 불편하게 사용하라는 코틀린 개발진들의 의도가 들어가있다고 합니다 😄 코틀린이 제공하는 기본 함수에 이를 사용하려고 한다면 non-nullable한 리턴값을 갖는 동일한 함수가 있는지 확인해봐도 좋을 것 같아요.

_Originally posted by @ddaaac in https://github.com/next-step/kotlin-racingcar/pull/760#discussion_r1015689624_

giyeon95 committed:
refactor: 예외값 input시 예외 발생에 대한 테스트 케이스 추가

예외값에 대한 케이스도 정의해 두셨는데, 이에 대한 테스트도 있으면 좋을 것 같아요.

_Originally posted by @ddaaac in https://github.com/next-step/kotlin-racingcar/pull/760#discussion_r1015690957_

giyeon95 commented:
리뷰주신 내용 학습 및 수정 후 다시 리뷰요청 드립니다.
 kotlin에서 지원하는 다양한 기능이 많이 있네요.👍

위에 리뷰주신 내용중 몇가지 질문도 함께 작성을 해두었는데 한번 같이 확인 부탁드립니다.🙏

ddaaac commented on src/test/kotlin/racingcar/RaceStageTest.kt:
```diff
@@ -0,0 +1,28 @@
+    @Test
+    @DisplayName("자동차가 갯수만큼 생성된다.")
+    internal fun createTest() {
+        // given
+
+        // when
+        val raceStage = RaceStage.create(5)
```
굳이 나누자면 given이 5가 되어야하지 않을까요 😄

ddaaac reviewed (approved):
안녕하세요 기연님 👋 

리뷰 반영 너무 잘 해주셨습니다 💯 
다음 단계 진행해주세요!

남겨주신 코멘트에 대한 답변 달았으니 다음 단계 진행하실 때 확인 부탁드려요.
